File: C++/1038-recover_the_smallest_number.cc
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <numeric>


bool comp(const std::string &n1, const std::string &n2) {
    return n1 + n2 < n2 + n1;
}


std::string
recover_the_smallest_num(std::vector<std::string> &nums)
{
    std::sort(nums.begin(), nums.end(), comp);
    std::string smallest = std::accumulate(
        nums.begin(), nums.end(), std::string());
    auto bgn_iter = std::find_if(smallest.begin(), smallest.end(),
                                 [](const char &c) {return c!='0';});
    if (bgn_iter != smallest.end()) {
        smallest.erase(smallest.begin(), bgn_iter);
    } else {
        smallest = "0";
    }
    return smallest;
}


int main() {

    int N;
    std::cin >> N;

    std::vector<std::string> nums;
    while (N--) {
        std::string num;
        std::cin >> num;
        nums.push_back(num);
    }

    std::cout << recover_the_smallest_num(nums) << '\n';

    return 0;
}

File: Python/1038-recover_the_smallest_number.py
from functools import cmp_to_key


def recover_the_smallest_num(nums):
    nums.sort(key=cmp_to_key(lambda x, y: -1 if x + y < y + x else 1))
    smallest = ''.join(nums).lstrip('0')
    if smallest:
        return smallest
    else:
        return '0'


if __name__ == '__main__':
    nums = input().split()[1:]
    print(recover_the_smallest_num(nums))
